[KEP-2395] Update metadata for 2395-removing-in-tree-cloud-providers

elmiko opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description:

Update version numbers to reflect current state of work for KEP-2395.

<!-- link to the k/enhancements issue -->
- Issue link:

<!-- other comments or additional information -->
- Other comments:

elmiko commented:
not quite sure why the test is failing, this error is unfamiliar to me:

```
 time="2023-09-06T18:56:23Z" level=info msg="the following PRR validation errors occurred in /home/prow/go/src/github.com/kubernetes/enhancements/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/kep.yaml:"
time="2023-09-06T18:56:23Z" level=info msg="getting PRR approver for beta stage: the selected PRR milestone stage is not populated"
    metadata_test.go:39: 
        	Error Trace:	metadata_test.go:39
        	Error:      	"map[/home/prow/go/src/github.com/kubernetes/enhancements/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/kep.yaml:[getting PRR approver for beta stage: the selected PRR milestone stage is not populated]]" should have 0 item(s), but has 1
        	Test:       	TestValidation
--- FAIL: TestValidation (0.57s) 
```

will continue to investigate

elmiko commented:
looking through the test code, i'm not sure if i'm missing a field in the `kep.yaml` or if we need an approver to move to beta stage.

elmiko commented:
as discussed in the sig cloud provider meeting last week, i am going to review the template and update the instructions for migration as well on this pr.

adding a hold while i am still wip

/hold

elmiko commented:
updated to new template, i have added "TBD" in the new sections. in the next PRs i will adjust some of the old content to align with the new sections, and then start to add more content for the beta phase.

elmiko commented:
updated to add a PRR approver for beta stage.

@wojtek-t i added you as the approver, if that is not appropriate i am happy to add someone else.

wojtek-t commented:
> @wojtek-t i added you as the approver, if that is not appropriate i am happy to add someone else.

Queued.
You may most probably expect some shadow reviewer making a pass before me though.

Also - can you please work with the SIG to have it approved from the SIG pov and have it added to the tracking board that we use also in PRR (by opting-in the corresponding feature)?

elmiko commented:
thanks @wojtek-t 

> Also - can you please work with the SIG to have it approved from the SIG pov and have it added to the tracking board that we use also in PRR (by opting-in the corresponding feature)?

yes, we are currently working through this, i have started adding labels to the issue for inclusion on the 1.29 tracking board.

elmiko commented:
updated to fill out more sections of the template. i've moved some of the previous "Approach" section to be part of the "Proposal" section, and then moved the phase specific details to the new "Design Details" section. i've also added some language about risks and mitigations, and updated the implementation history.

enj commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,6 +236,191 @@ import (
+
 ## Production Readiness Review Questionnaire
 
 ### Feature Enablement and Rollback
```
On `Can the feature be disabled once it has been enabled (i.e. can we roll back
the enablement)?`, what efforts have been taken to make sure the new code doesn't make changes that would cause the old code to misbehave?  Do we have any tests that try to exercise this fallback?

enj replied:
The `Rollout, Upgrade and Rollback Planning` section need to be filled out.

Same for `Monitoring Requirements`, `Dependencies`, `Troubleshooting`.

enj replied:
The `Scalability` section isn't correct.  For example this answer:

> No, it will actually remove calls to the cloud provider in all core components.

misses the point of the question completely.  We aren't asking in relation to who is making the calls, just that the calls are being made.  So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

elmiko replied:
>  what efforts have been taken to make sure the new code doesn't make changes that would cause the old code to misbehave?

if i understand your question correctly, the e2e tests that are currently in place will ensure that the old code is not active.

>  Do we have any tests that try to exercise this fallback?

i do not think we have any testing which exercises the rollback. i'll have to talk with the sig more about this, but this testing will be extremely difficult as it requires the core controllers to be reconfigured and then restarted. i don't think we had scoped this type of testing, i will double check though.

> The Rollout, Upgrade and Rollback Planning section need to be filled out.
> Same for Monitoring Requirements, Dependencies, Troubleshooting.

i'm still working on these, i left a note in my latest comment but i'd like to get more help from the sig answering these.

> misses the point of the question completely. We aren't asking in relation to who is making the calls, just that the calls are being made. So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

i did not write this answer, but the answer is factual from the core component perspective. this change will remove many calls to cloud providers that currently exist in the kubernetes code. but, it also changes that focus to the external cloud provider implementations.

in terms of the effort that has been taken to ensure we aren't creating extra load, that effort is being moved to the external CCMs which will be maintained by their individual communities. the common cloud-provider framework that the sig maintains does make improvements to the way the cloud controllers batch calls to the providers, but each CCM project is free to choose their implementation. 

one thing is true though, there is no way for the new code to create more load on the cloud provider APIs than the old code as the new code is a removal of all the old code. i suppose the confirmation of this is that tests which require cloud controllers simply will not work if there are no external CCMs engaged, there is no cloud to call as none of that code exists without the external CCM running.

elmiko replied:
> That is not what I mean. Say a field in a resource was always set to 5 with the legacy code, and always set to 6 by the new code. On a downgrade, the resource would still exist, but now the old code would see a 6, which it may not understand or act correctly against.

i'm having some trouble following this example. i'd like to talk through it.

in the case of this KEP, if a user upgrades to the new behavior, they will be disabling some loops that run in kubelet/kas/kcm and running them as external pods provided by their cloud infrastructure provider (or whichever team is maintaining, but not core kubernetes). if they  upgrade then they are changing flags in kubelet/kas/kcm and adding pods to the cluster, if they downgrade then they are re-enabling those flags and terminating the pods they were running. following from your example, let's say that the in-tree (old code) modifies a Node object with some value (the theoretical `5`), then the user upgrade to external (new code) and the value get modified (to a `6`), finally they downgrade back to the original code and now there is an error.  am i getting closer to your intention?

afaik, the former in-tree cloud controllers which became the external CCMs are not modifying Node or Service objects in a way that would create this incompatibility, but it is certainly possible. our reliance here has been on the behavioral tests in the e2e suite to confirm that Kubernetes continues to operate as expected. with that said though, due to the nature of testing specific cloud providers on their own infrastructure this testing has always been limited.

> I don't see how this can be true. A new controller could easily have a hot loop bug that the old controller did not. So by upgrading, the new controller would create way more load against the cloud provider than before, even if everything works correctly in CI.

perhaps i'm just misunderstanding, part of my disconnect here is that this KEP describes a process where users will need to source an external CCM for their clusters depending on the cloud provider they are running on. it is certainly possible that a user could run a controller which hot loops and creates extra load but i don't think it's in scope for this KEP to cover all the possible controllers that could be created. when this KEP was in alpha state, all the code from the in-tree controllers was moved into out-of-tree repositories that are maintained by provider specific teams. it is possible that in the interim those projects have added more to their controller loops, but as i understand our test infrastructure we have not been testing previous controllers on all the providers due to resource constraints (e.g. donated hardware/cloud time). given the testing that is currently done for in-tree cloud controllers, i'm not sure how we would answer the call to test all the possible controllers or even just test the controllers that were previously in-tree. 

fwiw, we (sig cloud provider) are working towards improving the testing situation for all cloud providers, but as i understand it currently we only deeply test the cloud controllers on gce, with a subset of those tests running on aws, and a subset of those running on azure. in the future we will have a testing suite that providers can run from their external CCM repositories to more deeply test any provider who wishes to participate.

back to your original question:

> So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

we have just enabled the external CCMs by default in the kubernetes CI, i'm not sure we have specific tests there which show the differences in load but i will ask around to see if there is something we could use as a reference. if not, we will need to develop those tests.

elmiko replied:
@enj i updated the answers to the api call questions to better reflect what is happening and might happen in the future. does this help?

enj reviewed (changes requested):
PRR shadow review.  Please update this KEP so that all sections are filled in, otherwise it is impossible to review.

elmiko commented:
bunch of updates from our sig meeting last week, i am still working on finishing the production readiness review questionnaire, but i think i will need some help from the sig to answer those.

i am unholding as i think we are good shape for more reviews, and i will update again after our sig meeting on wednesday.

cc @andrewsykim , @cheftako , @bridgetkromhout ptal

/hold cancel

elmiko commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -123,7 +164,7 @@ For alpha, the feature gates will be used for testing purposes. When enabled, te
-choose to continue using the in-tree provider by explicitly disabling the feature gates. Beta is targeted for v1.26 with the caveat that a majority of our CI signal jobs across providers should have converted to use CCM by then.
+choose to continue using the in-tree provider by explicitly disabling the feature gates. Beta is targeted for v1.29 with the caveat that a majority of our CI signal jobs across providers should have converted to use CCM by then.
 
 For GA, the feature gate will be enabled by default and locked. Users at this point MUST migrate to external components and use of the in-tree cloud providers will be disallowed. GA is targeted for v1.27. One release after GA, the in-tree cloud providers can be safely removed. 
```
i made this say 1.30, although i think that is aggressive at least it's in the future XD

elmiko commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,6 +236,191 @@ import (
+
+- <test>: <link to test coverage>
+
+TBD
```
i've already updated this, you were looking at an older copy.

elmiko commented on make.out:
done, thanks!

elmiko commented:
updated production readiness review questionnaire

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,6 +242,176 @@ import (
+
+- https://github.com/kubernetes/kubernetes/pull/117503
+
+The following test grids show the e2e tests running with external CCMS on GCP
```
nit what is CCMS?

elmiko replied:
oops, should be CCMs

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,6 +242,176 @@ import (
+which enable external CCMs by default for all GCE/GCP testing from the Kubernetes
+repository:
+
+- https://github.com/kubernetes/kubernetes/pull/117503
```
Hmm - why did we merge https://github.com/kubernetes/kubernetes/pull/117503
without having the KEP approved for beta?

@dims 

[Depending on what I will see on the PRR itself, I would say a possible outcome might be a need to revert that change... ?]

andrewsykim replied:
In @dims defense, we wanted the CI signal in really early in the release so we can establish a stable CI signal with all the in-tree cloud providers disabled before any other feature work started in 1.29. Agreed we can revert if all the PRR concerns aren't addressed, the good thing now is that we can revert the feature gate flip without having to revert all the CI jobs to use in-tree providers again.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -235,51 +452,101 @@ _This section must be completed when targeting beta graduation to a release._
 
 * **How can a rollout fail? Can it impact already running workloads?**
```
nit: you should update to the newer KEP template

elmiko replied:
i thought i did update to the newer template, is this a stale question?

elmiko replied:
i see that i did not fully copy the template, i'm updating now.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -235,51 +452,101 @@ _This section must be completed when targeting beta graduation to a release._
 * **How can a rollout fail? Can it impact already running workloads?**
 
-  TBD for beta.
+The primary method of rollout failure is due to configuration issues. When
```
> Can the feature be disabled once it has been enabled (i.e. can we roll back
  the enablement)

That section should at least mention the guide that descibed how upgrades/downgrades should be done
[already mentioned somewhere above]


> Are there any tests for feature enablement/disablement?

Were they added? Please update this answer.

elmiko replied:
ack, will update

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -235,51 +452,101 @@ _This section must be completed when targeting beta graduation to a release._
 * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
 
-  TBD for beta.
+  * Yes, tested manually by various providers, notably AWS and GCE.
```
Can you please be more specific? Any links?

elmiko replied:
i don't think we have links, but i will ask.

elmiko replied:
added a little more here, let me know if that works.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -235,51 +452,101 @@ _This section must be completed when targeting beta graduation to a release._
 the health of the service?**
 
-  TBD for beta.
+  * Pods scheduling, if the CCM node controller is not working properly the
```
pod scheduling often doesn't require CCM...

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -235,51 +452,101 @@ _This section must be completed when targeting beta graduation to a release._
+  * Nodes being initialized properly, zone and region labels being added.
+  * Pods being scheduled.
+  * Load balancer type services being created and destroyed.
+  * It is difficult to be more proscriptive about the SLOs of individual CCMs
```
So let's change to something:

"The existing SLOs for functionalities X, Y and Z should be used - the change is supposed to be no-op from end-user perspective."

elmiko replied:
ack, will update.

wojtek-t reviewed (commented):
I just took a quick pass - @enj 's concerns are the ones that I would bring up too, I just added few more.

elmiko commented:
> I just took a quick pass - @enj 's concerns are the ones that I would bring up too, I just added few more.

what about my last response to @enj ?

i've talked with the SIG about my answers and received general support there, i'm confused about what more detail i can add to satisfy the questions. i'd be happy to meet on video chat if it would be helpful to explain these points.

elmiko commented:
updated with requested changes, also fixed up the template. i have a feeling we are still missing some contextual updates to answer the larger outstanding questions. could use some assistance from @andrewsykim based on our discussions from the last sig meetings.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
 
-* **How does this feature react if the API server and/or etcd is unavailable?**
+###### Are there any tests for feature enablement/disablement?
+Adequate unit tests, component integration test and e2e tests will be added for this feature before
```
This was promoted to beta in k/k in https://github.com/kubernetes/kubernetes/pull/117503

Can you please specify what exact tests exist [it says they "will be added [...] before beta" so everything should already be there.

elmiko replied:
that's a great point, i think i'll need to confer with the sig about this. this part was written before i added the new content so i'll need to figure out which tests they were referring to.

elmiko replied:
i have added the enumerated unit tests from the external CCM framework, and linked to the PR which enables them in CI.

should i add a mention of no specific testing of a live upgrade/downgrade?

andrewsykim replied:
We just overhauled all existing kubernetes CI jobs to now run with the external GCE cloud provider, that is efffectively the test coverage. That plus additional test suites run by each provider downstream. 

We don't necessarily need to link all the test specifically, but maybe add some links to testgrid dashboards if we haven't already. 

We don't have any automated tests that enables and disables this in a test though, for the same reasons we don't have version upgrade/downgrade tests. But as mentioned in other threads, we have signal  from various manual tests and providers who have implemented the the migration downstream (GKE, EKS, AKS, etc)

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
 
-* **What are other known failure modes?**
+###### How can an operator determine if the feature is in use by workloads?
+* In genreal, if pods are being scheduled to the cluster then this feature is
```
IIUC you wanted to capture the fact that credential provider works. If so, let's be more explicit that what we are looking for is successful fetches of container images.

elmiko replied:
i've changed the wording on this one to make it more explicit.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+* Nodes being initialized properly, zone and region labels being added.
+* Pods being scheduled.
```
Same here - let's talk about pulling images, not scheduling pods.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
+* Pods being scheduled.
+* Load balancer type services being created and destroyed.
+* The existing SLOs for functionalities the `kube-controller-manager` should be used,
+  this change is supposed to be no-op from the end-user perspective.
```
SLOs are about metrics (and their thresholds) - this is too hand-wavy answer for the question.

Please either be more specific, or just leave the last bullet point...

elmiko replied:
i've replaced the vague statements with something more concrete that references metric values.

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
+* The existing SLOs for functionalities the `kube-controller-manager` should be used,
+  this change is supposed to be no-op from the end-user perspective.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
```
SLIs are effectively metrics - please specify the list of metrics, otherwise it's not helpful for operators.

elmiko replied:
updated with specific metrics

wojtek-t commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -195,153 +242,387 @@ import (
+> kube-controller-manager, kube-scheduler, and cloud-controller-manager must not be newer than the kube-apiserver instances they communicate with. They are expected to match the kube-apiserver minor version, but may be up to one minor version older (to allow live upgrades).
 
-  No.
+## Production Readiness Review Questionnaire
```
@enj - are your concerns addressed by the recent updates?

I'm somewhat biased because we made this migration for GKE and we managed to make it work, so it's certainly possible and the observability was there.
But I somewhat agree that the safety mechanisms aren't perfect, OTOH I'm not entirely sure we have right foundations to build something with a decent ROI and blocking for the sake of blocking wouldn't make sense.

But I would like to hear your thoughts.

enj replied:
Yeah I think what we have here is good enough.  This change is incredibly large and having tests that prove full downgrade would be a large effort (but these things are good to think about and manually test at least once).  Having existing tests pass with the in-tree components swapped out serves as a good heuristic for overall functionality as well.

elmiko commented:
thank you @wojtek-t , makes sense to me. i'll do another cleanup pass.

elmiko commented:
updated to address review comments

wojtek-t commented:
@elmiko - from my perspective there are two things that are blocking me from approving PRR:
(1) https://github.com/kubernetes/enhancements/pull/4190/files#r1343920195
(2) clarifying if @enj is ok with https://github.com/kubernetes/enhancements/pull/4190/files#r1343938350

I will ping Mo for the second, but please work on (1) - the deadline is tomorrow...

enj commented:
> (2) clarifying if @enj is ok with https://github.com/kubernetes/enhancements/pull/4190/files#r1343938350
> 
> I will ping Mo for the second, but please work on (1) - the deadline is tomorrow...

This LGTM: https://github.com/kubernetes/enhancements/pull/4190/files#r1346425348

elmiko commented:
added a little more detail about tests

andrewsykim commented on keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md:
```diff
@@ -439,8 +439,26 @@ apiserver network proxy do not have to be stopped on rollback.
 All capabilities from in-tree cloud providers will be re-disabled.
 
 ###### Are there any tests for feature enablement/disablement?
```
Also mention here that we've updated all existing CI jobs to now run with external providers. This means every (or at least "most") e2e tests can pass on a cluster with in-tree cloud provider disabled. 

Some testgrid links to reference:
* https://testgrid.k8s.io/provider-gcp-periodics#disable-cloud-provider -- this was the first periodic job that ran with cloud provider disabled which we had for several releases now)
* https://testgrid.k8s.io/provider-gcp-periodics#E2E%20Full%20-%20Cloud%20Provider%20GCP%20-%20with%20latest%20k8s.io/kubernetes -- this ensures any breaking changes to k8s.io/cloud-provider are caught early 
* https://testgrid.k8s.io/google-gce#pull-kubernetes-e2e-gce -- example presubmit required for every PR that runs with in-tree cloud providers disabled

wojtek-t replied:
These are rather feature test then enablement/disablement, but I agree that helps too.

That being said, I sugggest switching this answer to something like:

We have a huge number of feature tests, including;
- unit tests under https://github.com/kubernetes/cloud-provider/blob/release-1.28
- e2e test suites running with in-tree providers disabled:
    https://testgrid.k8s.io/provider-gcp-periodics#disable-cloud-provider
    https://testgrid.k8s.io/provider-gcp-periodics#E2E%20Full%20-%20Cloud%20Provider%20GCP%20-%20with%20latest%20k8s.io/kubernetes

For enablement/disablement tests, we are relying on manul tests that were done downstream on at least two large cloud providers: GCP and AWS.

elmiko replied:
thanks! updating

elmiko commented:
updated to clarify enablement/disablement tests

wojtek-t commented:
/label tide/merge-method-squash

/approve PRR

Thanks a lot for bearing with me

andrewsykim commented:
LGTM! Thank you @elmiko for driving the KEP changes!
